### PR TITLE
Add a parameter list to EnumElementDecl

### PR DESCRIPTION
**Pull request by CodaFi**
Factors out the pure refactoring part of #15418.  It should be NFC but for some richer SourceKit info and the lifting of [a regression](https://github.com/apple/swift/commit/b5828e6a61f587387b99620b55321c77bf5c87c8#diff-a53bf0219e57890224cd32d1c4dc9082L113) because of a mis-applied semantic check. 

This models, but does not plumb through, default arguments.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

File: lib/Parse/ParsePattern.cpp [hunk 1]
```diff
@@ -56,9 +56,9 @@ static DefaultArgumentKind getDefaultArgKind(Expr *init) {
   llvm_unreachable("Unhandled MagicIdentifierLiteralExpr in switch.");
 }
 
-void Parser::DefaultArgumentInfo::setFunctionContext(AbstractFunctionDecl *AFD){
+void Parser::DefaultArgumentInfo::setFunctionContext(DeclContext *DC, MutableArrayRef<ParameterList *> paramList){
```
Comment (slavapestov) [inline lib/Parse/ParsePattern.cpp]: Line too long

File: test/attr/attr_autoclosure.swift [hunk 2]
```diff
@@ -124,7 +124,7 @@ class TestFunc12 {
 
 
 enum AutoclosureFailableOf<T> {
-  case Success(@autoclosure () -> T)  // expected-error {{@autoclosure may only be used on parameters}}
+  case Success(@autoclosure () -> T)
```
Comment (slavapestov) [inline test/attr/attr_autoclosure.swift]: Can you add a test that a use of this type checks?

**Review by slavapestov — APPROVED** [review]
Looks good

**Comment by CodaFi** [comment]
Bother, now I remember why I redid duplicate checking in the last patch.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
[Source compatibility failure](https://ci.swift.org/job/swift-PR-source-compat-suite/953/) is unrelated to my changes (seems related to partial metadata @rjmccall).

⛵️

**Comment by rintaro** [comment]
Now the valid "parameter" grammar are:
```
parameter(func|init) = name name? ':' type ( '...' | '=' expression )?
parameter(subscript) = name name? ':' type '...'?
parameter(operator)  = '_'? name ':' type ('=' expression )?
parameter(closure)   = '_'? name (':' type)?
parameter(enumcase)  = ('_'? name ':')? type
--
name = identifier | '_'
```
And several migration supports.
It's a bit of a mess.
@nkcsgexi and @harlanhaskins should be aware of this.

**Comment by CodaFi** [comment]
Harlan and I were just discussing this.  For now, the extended grammar is banned by diagnostic.

**Pull request merged**